numpy/random/mtrand/distributions.c

```diff
@@ -718,12 +719,33 @@ double rk_wald(rk_state *state, double mean, double scale)
     }
 }
 
+static void kb_interrupt_handler(int signum) {
+    /* signal handlers are extremely limited
+     * and attempting to raise a Python
+     * KeyboardInterrupt via the Python
+     * C API here causes a segfault
+     */
+    printf("\nCaught Keyboard Interrupt\n");
+    exit(1);
+}
+
 long rk_zipf(rk_state *state, double a)
 {
     double am1, b;
 
     am1 = a - 1.0;
     b = pow(2.0, am1);
+
+    /* using PyErr_CheckSignals() from
+     * the Python C API would allow
+     * for keyboard interrupt handling
+     * BUT results in a segfault
+     * after Ctrl+C, so an alternative
+     * approach is used here;
+     * Related to Issue #9829
+     */
+    signal(SIGINT, kb_interrupt_handler);
```

Comment:
The handler is installed on every call to rk_zipf? When do you remove it?

Reply:
It's in the wrong place anyway. The problem is that the while loop never exits because the acceptance-rejection based implementation achieves near 100% rejection for parameters near one on account of an unfortunate scaling.